### PR TITLE
Add win32 events alternative to unix signals

### PR DESCRIPTION
**Pull request by ashie**
**Which issue(s) this PR fixes**: 
None

**What this PR does / why we need it**: 
Since Windows doesn't have UNIX signals, features realized by them such as flushing buffers aren't implemented on Windows.
This PR adds triggers for them by using win32 events. Currently only supervisor process supports them.

In addition, this PR adds a new command `fluent-ctl`. It sends a simple command to a fluentd process.
The main purpose of this command is that providing a kill command alternative on Windows.
On UNIX, it just only send a UNIX signal to a specified process.

```
Usage: fluent-ctl COMMAND PID_OR_SIGNAME
    
Commands:
  shutdown
  restart
  flush
  reload
```

SIGNAME can be specified by fluentd's "-x" or "--signame" option.
Note that this command doesn't support sending events to windows service.
Probably we need to trigger these events via `sc` command for windows service.

This PR also fixes a bug that the graceful shutdown feature doesn't work correctly when Fluentd is launched as Windows service. On this case, worker processes were killed forcedly before receiving `GRACEFUL_STOP` command.

**Docs Changes**:
Add `fluent-ctl` to https://docs.fluentd.org/deployment/command-line-option

**Release Note**: 
Same as title.

### CODE REVIEW

File: bin/fluentd-windows-cmd [hunk 1]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env ruby
+
+require 'win32/event'
+require 'optparse'
+
+$LOAD_PATH.unshift(File.join(__dir__, 'lib'))
+require 'fluent/command/windows_cmd'
+
+Fluent::WindowsCmd.new.call
```
Comment (cosmo0920) [inline bin/fluentd-windows-cmd:9]: Should we raise an error when using non-Windows environment?

File: bin/fluentd-windows-cmd [hunk 2]
```diff
@@ -0,0 +1,9 @@
+#!/usr/bin/env ruby
+
+require 'win32/event'
```
Comment (cosmo0920) [inline bin/fluentd-windows-cmd:3]: On non-Windows environment, `win32/event` shouldn't be installed as a dependency and loaded.
Should we use `Fluent.windows?` for guarding to require on non-Windows environment?

Comment (cosmo0920) [reply bin/fluentd-windows-cmd:3]: Or, just raise LoadError here is enough?

File: bin/fluentd-windows-cmd (hunk 1 above)
Comment (ashie) [reply bin/fluentd-windows-cmd]: I renamed the command to fluentdctl and support also UNIX environment, it just only sends a UNIX signal to a specified process though.

File: bin/fluentd-windows-cmd (hunk 2 above)
Comment (ashie) [reply bin/fluentd-windows-cmd]: Ditto

**Comment by ashie** [comment]
Matters to be discussed:

* I'm not sure the command name `fluentdctl` is appropriate or not.
  * `fluentd-ctl`? `fluentctl`? `fluent-ctl`?
* Although I've simplified the command names ASAP, I'm not sure they are appropriate or not.
  * e.g.) omit `buffers` of `flush` command, omit `graceful` of `reload` command
* UNIX version of `fluentdctl` may not be needed.
  * It just only send a UNIX signal to a specified process.

**Comment by cosmo0920** [comment]
I'd confirmed that `fluentdctl` can work on both of Windows and Linux.

> * I'm not sure the command name `fluentdctl` is appropriate or not.
>    * `fluentd-ctl`? `fluentctl`? `fluent-ctl`?

For current command naming convention, I vote `fluentctl` or `fluent-ctl`.

**Comment by ashie** [comment]
> For current command naming convention, I vote `fluentctl` or `fluent-ctl`.

I've renamed it to fluent-ctl.

**Comment by ashie** [comment]
Now I've found that both the original implementation for shutdown and this PR doesn't support multiple workers #3066.

**Comment by ashie** [comment]
Now I''ve found that fluent-ctl can't send events when fluentd is run as windows service, even if fluent-ctl is run as "SYSTEM" user.
So that I renamed the 3rd argument of the command from `PID_OR_SVCNAME` to `PID_OR_SIGNAME`, and removed the default parameter for it.

For Windows service, probably we need to trigger the events via `sc` command.

**Comment by ashie** [comment]
> Now I've found that both the original implementation for shutdown and this PR don't support multiple workers: #3066

It's ServerEngine's fault, I'll fix it. This PR should follow the change.

**Comment by ashie** [comment]
> Now I've found that fluent-ctl can't find events when fluentd is run as windows service, even if fluent-ctl is run as "SYSTEM" user.

Windows Service processes are separated as "Session 0" as of Windows Vista.
[PsExec](https://docs.microsoft.com/en-us/sysinternals/downloads/psexec) has a option to execute a command on "Session 0".
Now I've confirmed that I can trigger the events by the following command
```
\path\to\psexec.exe -s -i 0 \path\to\fluent-ctl restart fluentdwinsvc
```

**Comment by ashie** [comment]
> This PR also fixes a bug that the graceful shutdown feature doesn't work correctly when Fluentd is launched as Windows service.

I've filed a new pull request to fix it: #3156

**Comment by ashie** [comment]
The remaining issue of this PR is:

> SIGNAME can be specified by fluentd's "-x" or "--signame" option.
> Note that this command doesn't support sending events to windows service.
> Probably we need to trigger these events via sc command for windows service.

I'm now working on it.
`sc` command can send pre-defined control codes or user-defined control codes to a service:

https://docs.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-controlservice

* For `USER2` signal (gaceful-reload feature) on UNIX, probably mapping it to `SERVICE_CONTROL_PARAMCHANGE`  is suitable.
* For `USER1` (flush) and `HUP` (restart), there is no suitable pre-defined control code, so I'll map these features to user-defined control codes.
  * But [win32-sercvice](https://github.com/chef/win32-service/blob/master/lib/win32/daemon.rb) doesn't support receiving user-defined control codes. I'll send PR to it to enable receiving them.
  * Or we can map these features to rarely used control codes such as `SERVICE_CONTROL_NETBIND*` (deprecated control codes), but it's not so good design.

**Comment by ashie** [comment]
>  * But [win32-sercvice](https://github.com/chef/win32-service/blob/master/lib/win32/daemon.rb) doesn't support receiving user-defined control codes. I'll send PR to it to enable receiving them.

https://github.com/chef/win32-service/pull/74

**Comment by ashie** [comment]
And I've pushed the patch to send /receive a control code to/from fluend windows service.

shutdown:
```
# sc control 1 fluentdwinsvc
or
# sc control stop fluentdwinsvc
or
# fluent-ctl shutdown
```

reload:
```
# sc control paramchange fluentdwinsvc
or
# sc control  6 fluentdwinsvc
or
# fluent-ctl reload 
```

restart (need the patch for win32-service):
```
# sc control fluentdwinsvc 128
or
# fluent-ctl restart
```

flush (need the patch for win32-service):
```
# sc control fluentdwinsvc 129
or
# fluent-ctl flush
```

**Comment by ashie** [comment]
I already completed this work but we need to wait next release of win32-service (probably 2.3.0).

File: lib/fluent/supervisor.rb [hunk 3]
```diff
@@ -159,37 +165,101 @@ def stop_counter_server
...
-          stop(true)
         ensure
-          ev.close
+          events.each { |event| event.close }
         end
       end
     end
 
+    def stop_windows_event_thread
+      if Fluent.windows?
```
Comment (repeatedly) [inline lib/fluent/supervisor.rb:256]: L73 is `stop_windows_event_thread if Fluent.windows?`.
Is this `if` for more safety?

File: lib/fluent/supervisor.rb [hunk 4]
```diff
@@ -745,33 +831,45 @@ def install_main_process_signal_handlers
...
+      end
+    end
 
-      if Fluent.windows?
-        command_pipe = STDIN.dup
-        STDIN.reopen(File::NULL, "rb")
-        command_pipe.binmode
-        command_pipe.sync = true
+    def install_main_process_command_handlers
+      command_pipe = $stdin.dup
```
Comment (repeatedly) [inline lib/fluent/supervisor.rb:848]: Why change from `STDIN` to `$stdin`?

Comment (ashie) [reply lib/fluent/supervisor.rb:848]: For testing incoming commands at workers. See `test_main_process_command_handlers`.

File: lib/fluent/command/ctl.rb [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+      }
+      WINSVC_CONTROL_CODE_MAP = {
+        shutdown: SERVICE_CONTROL_STOP,
+        # 128 - 255: user-defined control code
+        # See https://docs.microsoft.com/en-us/windows/win32/api/winsvc/nf-winsvc-controlservice
+        restart: 128,
+        flush: 129,
+        reload: SERVICE_CONTROL_PARAMCHANGE,
+      }
+      COMMAND_MAP = WINDOWS_EVENT_MAP
```
[the remark this answers is not in the document]
Comment (ashie) [reply lib/fluent/command/ctl.rb:48]: No particular reason. I'll remove them.

File: lib/fluent/supervisor.rb (hunk 3 above)
Comment (ashie) [reply lib/fluent/supervisor.rb:256]: > Is this if for more safety?

That's right.

**Review by repeatedly — APPROVED** [review]

File: test/test_supervisor.rb [hunk 6]
```diff
@@ -190,6 +248,38 @@ def test_rpc_server
...
+    end
+    server.rpc_endpoint = sys_conf.rpc_endpoint
+
+    server.run_rpc_server
+
+    mock(server).restart(true) { nil }
+    Net::HTTP.get URI.parse('http://127.0.0.1:24447/api/plugins.flushBuffers')
+
+    server.stop_rpc_server
+  end
```
Comment (cosmo0920) [inline test/test_supervisor.rb:281]: `Net::HTTP.get URI.parse('http://127.0.0.1:24447/api/plugins.flushBuffers')`'s return value should be checked with some assertions?

**Comment by cosmo0920** [comment]
win32-service 2.3.0 which includes user-defined control code handling has been released:
https://github.com/chef/win32-service/commit/bde5569a9512f2f7ae07a754ea29f4c7b7b1abd0

**Comment by cosmo0920** [comment]
We should bump up dependent win32-service gem version to 2.3.0 before merging into master:
https://github.com/fluent/fluentd/blob/master/fluentd.gemspec#L36

**Review by cosmo0920 — APPROVED** [review]
I also confirmed that `fluent-ctl` can work for `shutdown/restart/flush/reload` on Windows Service.

BTW, I feel that fluent-ctl start is missing for Windows Service....
Starting Windows Service should be executed from sc/net command or Windows Management tool's Service?

**Comment by ashie** [comment]
> Starting Windows Service should be executed from sc/net command or Windows Management tool's Service?

That's right.

**Comment by repeatedly** [comment]
If there is no more comments, I will merge this patch.

**Comment by cosmo0920** [comment]
> If there is no more comments, I will merge this patch.

:+1:

Bumping up dependent win32-service gem version to v2.3.0 should do later.
win32-service v2.3.0 is not released yet.

**Comment by cosmo0920** [comment]
Let's :shipit:

**Comment by cosmo0920** [comment]
:tada:

**Comment by chook** [comment]
this is amazing, what about SIGCONT or forcing a thread dump @ashie

**Pull request merged**